cmd/agent/app/reporter/grpc/reporter.go

```diff
@@ -71,3 +74,30 @@ func (r *Reporter) send(spans []*model.Span, process *model.Process) error {
 	}
 	return err
 }
+
+// addTags appends jaeger tags for the agent to every span it sends to the collector.
+func addProcessTags(spans []*model.Span, process *model.Process, agentTags []model.KeyValue) ([]*model.Span, *model.Process) {
+	if len(agentTags) == 0 {
+		return spans, process
+	}
+	if process != nil {
+		process.Tags = append(process.Tags, agentTags...)
+		return spans, process
+	}
+	for _, span := range spans {
+		// span.Process will never be nil -
+		// Zipkin batch will always populate span.Process
+		span.Process.Tags = append(span.Process.Tags, agentTags...)
+	}
```

Comment:
I don't think we should be hardcoding the assumption that `process` and `span.Process` are mutually exclusive. There are cases when a single application generates spans on behalf of different services (e.g. a routing sidecar). It is possible that a single submitted Batch has `batch.Process` defined, but also some individual `span.Process` are not nil. There's no reason to exclude this use case from this function, just need to (a) not return in L85, and (b) check that `span.Process != nil` before L90.

Reply:
Thanks for the explanation, the use-case was unclear to me, this was very helpful :)

Reply:
it's the general principle: don't assume the clients will behave in the expected way, try to make sense of all possible inputs